### ISSUE TITLE
spotty-custom periodically dumps core (FreeBSD)

### ISSUE THREAD

**Issue by hartzell** [michaelherger/spotty#18]
I'm using spotty on FreeBSD (11.2).   It generally works well, the only issue is that it occasionally stops playing music and resyncs.  I've never been sure whether it's a spotty problem, or a sync-ing-among-my-players problem, or a network problem, or bad karma.

Periodically my FreeBSD daily reports mention that spotty-custom has dumped a core file.  I found the time to go take a look at it today, but it seems that spotty custom isn't built with debugging info:

```
# gdb /var/db/logitechmediaserver/cache/InstalledPlugins/Plugins/SpottyBinFreeBSD/Bin/spotty-custom /var/tmp/spotty-custom.core
GNU gdb 6.1.1 [FreeBSD]
Copyright 2004 Free Software Foundation, Inc.
GDB is free software, covered by the GNU General Public License, and you are
welcome to change it and/or distribute copies of it under certain conditions.
Type "show copying" to see the conditions.
There is absolutely no warranty for GDB.  Type "show warranty" for details.
This GDB was configured as "amd64-marcel-freebsd"...Dwarf Error: wrong version in compilation unit header (is 4, should be 2) [in module /var/db/logitechmediaserver/cache/InstalledPlugins/Plugins/SpottyBinFreeBSD/Bin/spotty-custom]

Core was generated by `/var/db/logitechmediaserver/cache/InstalledPlugins/Plugins/SpottyBinFreeBSD/Bin/'.
Program terminated with signal 6, Aborted.
#0  0x000000080196847a in ?? ()
(gdb) where
#0  0x000000080196847a in ?? ()
#1  0x0000000801968444 in ?? ()
#2  0x00000000000188cf in ?? ()
#3  0xb60cd4abe43ff4b5 in ?? ()
#4  0x00007fffffffcb24 in ?? ()
#5  0x0000000000000000 in ?? ()
(gdb)
```

I'm not a rustacean; I'm not sure how easy it would be to provide a version with debugging symbols and/or whether it'd be likely to tell us anything useful.

If you could and if it would, I'd be game to run it for a while to see where it's giving up the ghost.

**Comment by hartzell** [comment]
Forgot the version info:

```
# /var/db/logitechmediaserver/cache/InstalledPlugins/Plugins/SpottyBinFreeBSD/Bin/spotty-custom -h
spotty v0.20.0
error: Unrecognized option: 'h'
Usage: /var/db/logitechmediaserver/cache/InstalledPlugins/Plugins/SpottyBinFreeBSD/Bin/spotty-custom [options]
[...]
```

LMS is built using [my hacked up version of the FreeBSD port](https://github.com/hartzell/freebsd-ports/tree/master/audio/logitechmediaserver), which pulls from a recent commit in Logitech/slimserver (Logitech:slimserver:f6c4ca2, Logitech:slimserver-vendor:8b05092:vendor).


```
Logitech Media Server Version: 7.9.2 - TRUNK @ UNKNOWN
Hostname: lms
Server IP Address: 192.168.42.242
Server HTTP Port Number: 9000
Operating system: FreeBSD - EN - utf8
Platform Architecture: amd64-freebsd
Perl Version: 5.30.0 - amd64-freebsd-thread-multi
Audio::Scan: 1.02
IO::Socket::SSL: 2.066
Database Version: DBD::SQLite 1.58 (sqlite 3.22.0)
Total Players Recognized: 3
```

**Comment by michaelherger** [comment]
You should try to get in touch with Richard Hermann (his email should be listed on the Plugins page). He's kind enough to provide the BSD binaries for Spotty.

**Comment by hartzell** [comment]
done, pointed him here for add'l info.

**Comment by hartzell** [comment]
I haven't heard back from Richard.  Turns out that compiling the binary is dead simple.  I set up a fresh FreeBSD 11.2 jail (I'm working on a FreeNAS box), but it should work in any recent release, jailed or not:

```
pkg install git rust
git clone https://github.com/michaelherger/spotty.git
cd spotty/
cargo build --release
target/release/spotty -h
```

I "installed" the updated helper by tracking down the `spotty-custom` binary in the jail that's running my LMS server, backing it up, and replacing it with updated binary.

There's probably a cleaner way to do it.  I suppose I could host my own plugin repository, riffing on Richard's site: https://www.neversimple.eu/LMS/repodata.xml

Let's see how well the new release works and I'll take it from there.

**Comment by NeverSimple** [comment]
'Dead simple', was what I was gonna tell you :) It really is, there's no magic going into compiling the binary for FreeBSD. I guess most FreeBSD users actually compile their own, but FreeNAS users may benefit from a 'point-and-click' install. And it's one thing less to do yourself... It is downloaded a couple of hundred times each month peaking at a couple of Gigs of data when a new spotty version is released. So it probably fills a need.

I'm no rustacean either, and I do see the same behaviour with the core dumps that you see, just never followed up on it. The last binary I compiled (3.0, I will build 3.5 today) , I used FreeBSD-12.1. Building 'spotty-custom' on this FreeBSD version works fine, but the result won't run on FreeBSD-11.x (at least not on 11.2; 'Undefined symbol "fstat@FBSD_1.5"'). Because FreeNAS is on FreeBSD 11.2 that is something to be aware of, I guess.

Oh and Github just plainly refuses to sent me ANY notification I subscribe to, so usually I'm a little 'out-of-sync'...
Good luck chasing down the core dumps!

**Comment by hartzell** [comment]
Still trying to track this down, currently trying to get useful debugging info from spotty's core files.

I sent the following to community@rust-lang.org and freebsd-questions@freebsd.org, hopefully I'll get a useful clue:

---
---

Hi Folks,

TL;DR

How can I build a rust application on an amd64 FreeBSD system so that
I can use a debugger to examine its core files?

Using a "dev" build, rust@1.38.0, `rust-gdb` and gdb@8.3.1 don't seem
to get me there.

---

I'm hoping for feedback on, or other places to ask, a question about
getting debugging info from core dumps from a rust application on
FreeBSD.

Here's the scenario: I listen to music using the Logitech Media Server
("Slimdevices") ecosystem
[https://en.wikipedia.org/wiki/Logitech_Media_Server].  There's a
plugin [https://github.com/michaelherger/spotty] that lets me listen
to spotify, via the librespot library.  The application periodically
crashes due to a SIGABRT.  I'm unable to get debug info from the
resulting core files.

I'm working on an amd64 FreeNAS system, which runs FreeBSD 11.2 under
the covers.  I've installed "everything" from packages; in particular

- `rustc --version` says: `rustc 1.38.0`
- the core gdb is 6.1.1
- the gdb from packages is 8.3.1.

I've cloned the application and built it with `cargo build`, which
seems to give me a "debugging binary".

```
root@spotty-build:~/spotty/target/debug # file ./spotty
./spotty: ELF 64-bit LSB shared object, x86-64, version 1 (FreeBSD), dynamically
 linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 11.2, FreeBSD-style, with
 debug_info, not stripped
root@spotty-build:~/spotty/target/debug #
```

I ran it by hand in the background and killed it via `kill -ABRT
<pid>` to generate a trivial core file so that I can figure out how to
get debugging symbols in the deployed application.  I think that this
is rust-related, but perhaps my gdb skills are just rusty.

- If I use the system gdb it seems to have trouble with version of the
  Dwarf debugging info:

    ```
    root@spotty-build:~/spotty/target/debug # gdb spotty /var/db/system/cores/spotty.core
    GNU gdb 6.1.1 [FreeBSD]
    Copyright 2004 Free Software Foundation, Inc.
    GDB is free software, covered by the GNU General Public License, and you are
    welcome to change it and/or distribute copies of it under certain conditions.
    Type "show copying" to see the conditions.
    There is absolutely no warranty for GDB.  Type "show warranty" for details.
    This GDB was configured as "amd64-marcel-freebsd"...Dwarf Error: wrong version in compilation unit header (is 4, should be 2) [in module /root/spotty/target/debug/spotty]

    Core was generated by `./spotty -n poodle'.
    Program terminated with signal 6, Aborted.
    #0  0x0000000803182b2c in ?? ()
    (gdb) where
    #0  0x0000000803182b2c in ?? ()
    #1  0x0000000803176bc1 in ?? ()
    #2  0x00007fffffffad60 in ?? ()
    #3  0x00000008031797cc in ?? ()
    #4  0x000000000001894b in ?? ()
    #5  0x0000000000000001 in ?? ()
    #6  0x00007fffffffadd0 in ?? ()
    #7  0x0000000803180858 in ?? ()
    #8  0x00007fffffffad90 in ?? ()
    #9  0x00000008040162b8 in ?? ()
    #10 0x0000000000000000 in ?? ()
    (gdb)
    ```

- gdb@8.3.1 (from packages) seems to handle the DWARF debugging info
  version but still no joy:

    ```
    root@spotty-build:~/spotty/target/debug # /usr/local/bin/gdb --version
    GNU gdb (GDB) 8.3.1 [GDB v8.3.1 for FreeBSD]
    Copyright (C) 2019 Free Software Foundation, Inc.
    License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.
    root@spotty-build:~/spotty/target/debug # /usr/local/bin/gdb spotty /var/db/system/cores/spotty.core
    GNU gdb (GDB) 8.3.1 [GDB v8.3.1 for FreeBSD]
    Copyright (C) 2019 Free Software Foundation, Inc.
    License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.
    Type "show copying" and "show warranty" for details.
    This GDB was configured as "x86_64-portbld-freebsd11.3".
    Type "show configuration" for configuration details.
    For bug reporting instructions, please see:
    <http://www.gnu.org/software/gdb/bugs/>.
    Find the GDB manual and other documentation resources online at:
        <http://www.gnu.org/software/gdb/documentation/>.

    For help, type "help".
    Type "apropos word" to search for commands related to "word"...
    Reading symbols from spotty...
    [New LWP 100683]
    [New LWP 100623]
    [New LWP 100649]

    warning: Unexpected size of section `.reg-xstate/100683' in core file.
    Core was generated by `./spotty -n poodle'.
    Program terminated with signal SIGABRT, Aborted.

    warning: Unexpected size of section `.reg-xstate/100683' in core file.
    #0  0x0000000803182b2c in ?? ()
    [Current thread is 1 (LWP 100683)]
    warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
    of file /root/spotty/target/debug/spotty.
    Use `info auto-load python-scripts [REGEXP]' to list them.
    (gdb) where
    #0  0x0000000803182b2c in ?? ()
    #1  0x0000000803176bc1 in ?? ()
    #2  0x00007fffffffad60 in ?? ()
    #3  0x00000008031797cc in ?? ()
    #4  0x000000000001894b in ?? ()
    #5  0x0000000000000001 in ?? ()
    #6  0x00007fffffffadd0 in ?? ()
    #7  0x0000000803180858 in ?? ()
    #8  0x00007fffffffad90 in ?? ()
    #9  0x00000008040162b8 in ?? ()
    #10 0x0000000000000000 in ?? ()
    (gdb) quit
    root@spotty-build:~/spotty/target/debug #
    ```

- using `rust-gdb` (and setting the path so that it picks up the newer
  gdb) clears up the debug script issue, but I'm still not getting
  useful line numbers:

    ```
    root@spotty-build:~/spotty # rust-gdb target/debug/spotty /var/db/system/cores/spotty.core
    GNU gdb (GDB) 8.3.1 [GDB v8.3.1 for FreeBSD]
    Copyright (C) 2019 Free Software Foundation, Inc.
    License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.
    Type "show copying" and "show warranty" for details.
    This GDB was configured as "x86_64-portbld-freebsd11.3".
    Type "show configuration" for configuration details.
    For bug reporting instructions, please see:
    <http://www.gnu.org/software/gdb/bugs/>.
    Find the GDB manual and other documentation resources online at:
        <http://www.gnu.org/software/gdb/documentation/>.

    For help, type "help".
    Type "apropos word" to search for commands related to "word"...
    Reading symbols from target/debug/spotty...
    [New LWP 100683]
    [New LWP 100623]
    [New LWP 100649]

    warning: Unexpected size of section `.reg-xstate/100683' in core file.
    Core was generated by `./spotty -n poodle'.
    Program terminated with signal SIGABRT, Aborted.

    warning: Unexpected size of section `.reg-xstate/100683' in core file.
    #0  0x0000000803182b2c in ?? ()
    [Current thread is 1 (LWP 100683)]
    (gdb) where
    #0  0x0000000803182b2c in ?? ()
    #1  0x0000000803176bc1 in ?? ()
    #2  0x00007fffffffad60 in ?? ()
    #3  0x00000008031797cc in ?? ()
    #4  0x000000000001894b in ?? ()
    #5  0x0000000000000001 in ?? ()
    #6  0x00007fffffffadd0 in ?? ()
    #7  0x0000000803180858 in ?? ()
    #8  0x00007fffffffad90 in ?? ()
    #9  0x00000008040162b8 in ?? ()
    #10 0x0000000000000000 in ?? ()
    (gdb)
    ```


I'd appreciate any help, suggestions or ... on how to get useful
information about crashed from these core files.

Thanks!

g.

**Comment by michaelherger** [comment]
Are you still seeing this with the latest code? I did modify some of the error handling in 1.2.0.

**Comment by hartzell** [comment]
Michael Herger writes:
 > Are you still seeing this with the latest code? I did modify some
 > of the error handling in 1.2.0.

I haven't noticed it, but haven't updated that recently and my entire
music system has been packed away for about 5 months after a move and
very slow unpacking.

I'll update once I get my home set up and will let you know if there
are any problems.

Thanks for sharing your work!

g.

**Comment by michaelherger** [comment]
I'm going to close this issue. Please feel free to open a new report on https://github.com/michaelherger/librespot, where this application is living now.
